### PR TITLE
Support tensor.backward in dynamo

### PR DESCRIPTION
**Pull request by tugsbayasgalan**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #172126
* __->__ #169415

Following https://github.com/pytorch/pytorch/issues/167729, one natural extension is that we should able to trace through tensor.backward. When we do tensor.backward, we rewrite it as autograd.grad + torch.ops.inductor.accumulate_grad_ in dynamo. When there is no explicit parameters being passed in to backward, we just look at dynamo graph inputs and find all input parameters. This works because autograd.grad will just return None for the unused ones. 

The change looks quite big but there are several small/independent fixes:
1. Have to put allow_unused parameter to autograd.grad when we are automatically query-ing parameters because we are passing in all possible parameters that may or may not correspond to particular autograd.grad call 
2. We try to collect in-graph created parameters 
3. We use node proxies to deduplicate parameters that might be aliasing each other
4. If user passes in non-leaf parameter to backward, we graph break because generic .grad handling in dynamo is limited today. Will do as a follow up. 
2. If we are calling backward on a tensor with no grad fn, we should error. 

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @kadeng @chauhang @amjames @Lucaskabela @jataylo

### CODE REVIEW

File: torch/_dynamo/variables/tensor.py [hunk 1]
```diff
@@ -1008,14 +1008,80 @@ def wrap(i, sub_proxy):
...
 
-    def method_backward(self, *args, **kwargs):
-        unimplemented(
-            gb_type="Unsupported Tensor.backward() call",
-            context=f"call_method {self} backward {args} {kwargs}",
-            explanation="Dynamo currently does not support tracing `Tensor.backward()`.",
-            hints=[*graph_break_hints.FUNDAMENTAL],
+    def method_backward(
+        self, gradient=None, retain_graph=None, create_graph=None, inputs=None
+    ):
```
Comment (ezyang) [inline torch/_dynamo/variables/tensor.py]: Almost definitely want to config guard this for now

**Review by ezyang — CHANGES_REQUESTED** [review]
You need to make sure that the autograd graph the backward() call calls doesn't extrude "beyond" the compiled region. Concretely, construct a test where a non-leaf requires grad True input is passed in and then try to call backward().

File: torch/_dynamo/variables/tensor.py [hunk 2]
```diff
@@ -1008,14 +1009,89 @@ def wrap(i, sub_proxy):
...
+        if not config.trace_autograd_ops:
+            unimplemented(
+                gb_type="Unsupported Tensor.backward() call",
+                context=f"call_method {self} backward {gradient} {retain_graph} {create_graph} {inputs}",
+                explanation="Dynamo currently does not support tracing `Tensor.backward()`.",
+                hints=[*graph_break_hints.FUNDAMENTAL],
```
Comment (williamwen42) [inline torch/_dynamo/variables/tensor.py]: Remove `FUNDAMENTAL` - replace with suggestion to turn on your config?

**Review by williamwen42 — DISMISSED** [review]

File: benchmarks/dynamo/ci_expected_accuracy/inductor_huggingface_training.csv [hunk 3]
```diff
@@ -34,35 +34,35 @@ DistillGPT2,pass,7
 
 
 
-ElectraForCausalLM,pass,4
+ElectraForCausalLM,pass,3
```
Comment (anijain2305) [inline benchmarks/dynamo/ci_expected_accuracy/inductor_huggingface_training.csv]: Why do these change? Isn't the flag defaulted to False?

File: torch/_dynamo/variables/tensor.py [hunk 4]
```diff
@@ -1008,14 +1009,89 @@ def wrap(i, sub_proxy):
...
+        # simply don't get gradients (no error is raised)
+        grad_kwargs["allow_unused"] = VariableTracker.build(tx, True)
+
+        # Call autograd.grad
+        grad_args = [self, inputs]
+        if gradient is not None:
+            grad_args.append(gradient)
+
+        autograd_grad_fn = VariableTracker.build(tx, torch.autograd.grad)
+        grads_var = autograd_grad_fn.call_function(tx, grad_args, grad_kwargs)
```
Comment (anijain2305) [inline torch/_dynamo/variables/tensor.py:1250]: Dont we have to detach the inputs with require_grad=True before calling autograd.grad? Otherwise, the grads can flow twice.

File: benchmarks/dynamo/ci_expected_accuracy/inductor_huggingface_training.csv (hunk 3 above)
Comment (tugsbayasgalan) [reply benchmarks/dynamo/ci_expected_accuracy/inductor_huggingface_training.csv]: I enabled them by default. If default=False, we can't know what the issues are.

File: torch/_dynamo/variables/tensor.py (hunk 4 above)
Comment (tugsbayasgalan) [reply torch/_dynamo/variables/tensor.py:1250]: Hmm i think no? Because we make sure the compiled region is guarded enough that it only takes leaf tensors.

Comment (anijain2305) [reply torch/_dynamo/variables/tensor.py:1250]: > Because we make sure the compiled region is guarded enough that it only takes leaf tensors.

Where do we ensure that?

Comment (tugsbayasgalan) [reply torch/_dynamo/variables/tensor.py:1250]: I do it in the PR below it https://github.com/pytorch/pytorch/pull/169493 since backward impl depends on supporting autograd.grad

File: torch/_dynamo/variables/torch.py [hunk 5]
```diff
@@ -592,6 +592,16 @@ def handle_get_default_nowrap_functions(
 
         @register(torch.ops.inductor.accumulate_grad_.default)
         def handle_accumulate_grad_(self, tx: "InstructionTranslator", *args, **kwargs):
+            # Check if gradient is sparse - sparse gradients are not supported
```
Comment (tugsbayasgalan) [inline torch/_dynamo/variables/torch.py]: @ezyang  this showed up when i enabled by default, i could remove this since we are off by default now to minimize code change, lmk.

File: benchmarks/dynamo/ci_expected_accuracy/inductor_huggingface_training.csv (hunk 3 above)
Comment (tugsbayasgalan) [reply benchmarks/dynamo/ci_expected_accuracy/inductor_huggingface_training.csv]: Turned off by default

File: torch/_dynamo/variables/tensor.py (hunk 1 above)
Comment (tugsbayasgalan) [reply torch/_dynamo/variables/tensor.py]: Done

File: torch/_dynamo/polyfills/__init__.py [hunk 6]
```diff
@@ -98,8 +98,6 @@ def accumulate_grad(x, new_grad):
     new_grad_strided.copy_(new_grad)
     if x.grad is None:
         x.grad = new_grad_strided
-    elif torch.is_grad_enabled():
```
Comment (tugsbayasgalan) [inline torch/_dynamo/polyfills/__init__.py:107]: This returns a new tensor, which i don't think is right?

Comment (ezyang) [reply torch/_dynamo/polyfills/__init__.py:107]: Why do you think this is not right? If grad is enabled during accumulate grad, it means that double backwards is involved and we need to do a differentiable operation (while technically `add_` will work, doing an out of place addition will more likely work and it's better). The real problem is probably that you didn't `no_grad` in the context where `accumulate_grad` is called; you need to replicate the autograd engine behavior here.

Comment (tugsbayasgalan) [reply torch/_dynamo/polyfills/__init__.py:107]: Yep i now wrap the accumulate_grad logic with no_grad context and added a test. I think i misunderstood the failure.

File: torch/_dynamo/variables/tensor.py [hunk 7]
```diff
@@ -1072,18 +1074,158 @@ def wrap(i: Any, sub_proxy: torch.fx.Proxy) -> VariableTracker:
...
         return VariableTracker.build(tx, out)
 
+    def _collect_backward_inputs(self, vars_iter, error_on_non_leaf=False):
+        """
+        Collect unique leaf tensors from vars_iter for backward.
+
+        Only collects leaf tensors (no grad_fn). Non-leaf tensors are skipped
+        (or error if error_on_non_leaf=True) since the accumulate_grad polyfill
+        cannot handle .grad access on non-leaf tensors as Dynamo creates generic
+        GetAttrVariable (https://github.com/pytorch/pytorch/issues/171204).
```
Comment (ezyang) [inline torch/_dynamo/variables/tensor.py]: The real reason to not include non-leafs, is when you are automatically collecting inputs, you MUST NOT stop the gradients at non-leafs

File: torch/_dynamo/variables/tensor.py [hunk 8]
```diff
@@ -1072,18 +1074,158 @@ def wrap(i: Any, sub_proxy: torch.fx.Proxy) -> VariableTracker:
...
+                explanation="Dynamo currently does not support tracing `Tensor.backward()` when trace_autograd_ops is off.",
+                hints=["Set torch._dynamo.trace_autograd_ops=True"],
+            )
+
+        if not self.requires_grad and not self.has_grad_fn:
+            raise TorchRuntimeError(
+                "tensor does not require grad and does not have a grad_fn"
+            )
+
+        # Step 1: Collect leaf tensors to compute gradients for
```
Comment (ezyang) [inline torch/_dynamo/variables/tensor.py:1196]: What would be safer here is to verify that you actually have a input source corresponding to every AccumulateGrad grad fn that you flood fill from the outputs passed to grad.

Comment (ezyang) [reply torch/_dynamo/variables/tensor.py:1196]: The extrusion test below might suffice for this.

File: torch/_dynamo/variables/tensor.py [hunk 9]
```diff
@@ -1072,18 +1074,158 @@ def wrap(i: Any, sub_proxy: torch.fx.Proxy) -> VariableTracker:
...
+        auto_detect = inputs is None
+        if auto_detect:
+            # Sources can be either user inputs (params are included here)
+            # or parameters that are created in forward.
+            all_vars = chain(
+                tx.output.leaf_var_creation_order,
+                tx.output.input_source_to_var.values(),
```
Comment (ezyang) [inline torch/_dynamo/variables/tensor.py:1208]: We should check what happens when it turns out there are NO known leafs, which will give you a torch.autograd.grad call with no inputs.

Comment (ezyang) [reply torch/_dynamo/variables/tensor.py:1208]: This works, but only because torch.autograd.grad handling is very conservative. And it turns out it MUST be conservative, because see the convo at https://github.com/pytorch/pytorch/pull/169493/files#r2665611139 But it would be better NOT to ever generate illegal torch.autograd.grad call (where there are not enough grad inputs) in the first place, by doing extra checks here.

Comment (ezyang) [reply torch/_dynamo/variables/tensor.py:1208]: So you want to build a mapping from grad_fn to TensorVariable here. The main complication is AccumulateGrad, where you have to do the one-hop trick to get them (since they are not exposed directly on the VT).

**Review by ezyang — APPROVED** [review]
